### ISSUE TITLE
Some clarifications needed about #1939 and related ones

### ISSUE THREAD

**Issue by P5music** [mattgarrish/epub-revision#1948]
In
https://github.com/w3c/epub-specs/pull/1939/files/3c951205f45e242ed5334a365fb90fe95fe9effe

-the URL name is used also for partial URLs or url segments, if I am not wrong.
Is it correct to address that reference strings as URLs?
Examples:
../../image1.png
Is it a URL?

-the base URL is often referred to when dealing with some URL fragments that are "more complete" (but not the entire URL), like when it is obtained by concatenating partial path segments.
But it is not clear if it includes the root container folder in those cases where it exists in implementation (like mobydick/).
In those cases where that folder exists, and the check includes it in the path, the algorithm will fail to detect out-of-container URLs (false positives and negatives are both possible I think),
or going-too-much-up URLs.
So it should be explicitly indicated that the A folder has to be replaced with the folder name or vice-versa, or just it has not to be included at the beginning of the URL (just in case someone is tricked into thinking it).

-and a distinction between the out-of-container URL and going-too-much-up URL concepts should be considered explicitly.

-it is not clear if the algorithm is to be applied to absolute URLs (I suppose http:// URLs are likely to be found, and not file:///).
In that case is it able to just return a value that has not the example.org/A part inside? (= URL is accepted)
Maybe the URL parser is able to parse something like:
http://example.org/A/http://myepubresourceswebsite/resources/mobydick/bonusimages/hires_cover.jpg
yielding
http://myepubresourceswebsite/resources/mobydick/bonusimages/hires_cover.jpg
In fact, the algorithm seems to expect that the original scheme-host part is not present in the result (= URL is accepted).
but I am no expert of URL parsing APIs, so other RSs could find it strange too.

Regards

**Comment by rdeltour** [comment]
@P5music you'll find answers to most of your questions in the [URL standard](https://url.spec.whatwg.org). This standard is normatively referenced by EPUB, which means EPUB is using the definitions and logic defined there.
To understand URL-related conformance statements in EPUB, you need to understand the referenced bits of the URL standard.

> -the URL name is used also for partial URLs or url segments, if I am not wrong. Is it correct to address that reference strings as URLs? Examples: ../../image1.png Is it a URL?

It is a URL string, not a URL record. "URL" by itself means URL record.


> -the base URL is often referred to when dealing with some URL fragments that are "more complete" (but not the entire URL), like when it is obtained by concatenating partial path segments.

it is not a simple concatenation. How a base URL is combine with a string to make a full URL is defined by the URL parsing algorithm.

> But it is not clear if it includes the root container folder in those cases where it exists in implementation (like mobydick/). 

This is defined in EPUB RS spec, [URL of the Root Directory](https://www.w3.org/TR/epub-rs-33/#sec-container-iri).

> In those cases where that folder exists, and the check includes it in the path, the algorithm will fail to detect out-of-container URLs (false positives and negatives are both possible I think), or going-too-much-up URLs. So it should be explicitly indicated that the A folder has to be replaced with the folder name or vice-versa, or just it has not to be included at the beginning of the URL (just in case someone is tricked into thinking it).

I don't understand what you mean by going-too-much-up URLs.
If you'd like to clarify, please provide an example including the file and directory structure of a container, and concrete URL strings.

> 
> -and a distinction between the out-of-container URL and going-too-much-up URL concepts should be considered explicitly.

Can you define those? I'm not sure I understand how they're different.

> 
> -it is not clear if the algorithm is to be applied to absolute URLs (I suppose http:// URLs are likely to be found, and not file:///).

The algorithm in #1939 applies to all URL strings. Even absolute-URL strings.

> In that case is it able to just return a value that has not the example.org/A part inside? (= URL is accepted) Maybe the URL parser is able to parse something like: http://example.org/A/http://myepubresourceswebsite/resources/mobydick/bonusimages/hires_cover.jpg yielding http://myepubresourceswebsite/resources/mobydick/bonusimages/hires_cover.jpg In fact, the algorithm seems to expect that the original scheme-host part is not present in the result (= URL is accepted).
> but I am no expert of URL parsing APIs, so other RSs could find it strange too.

As you guessed this is defined in URL parsing.
Parsing the string "`http://myepubresourceswebsite/resources/mobydick/bonusimages/hires_cover.jpg`" with the URL `http://a.example.org/A/` as based returns the URL `http://myepubresourceswebsite/resources/mobydick/bonusimages/hires_cover.jpg`. So step 9 of the algorithm returns true (i.e. the URL string is conforming, a conformance checker will raise no warning).

**Comment by P5music** [comment]
@rdeltour

-Sorry I cannot find explanation or examples in that official document to understand how the URL is also pieces of URL.
Any direct link to part of the documents would be welcome.

-If I put this URL
<code>https://url.spec.whatwg.org/https://github.com/w3c/epub-specs/issues/1948</code>
in the browser address bar
I get "404 Not Found" error. 
It does not become 
<pre>https://github.com/w3c/epub-specs/issues/1948</pre>
It is not accepted or parsed.
I also tried 
<code>https://url.spec.whatwg.org/A/https://github.com/w3c/epub-specs/issues/1948</code>
with the same results.
There is something I do not grab here. Can you explain? What URL parsers do you mean? Where is the official rule in the document to do so?

-About the two kind of leaking URLs, they are
out-of-container URLs
and
going-too-much-up URLs
They are not the same because there are two kind of problems with URLs for the ePub3:

1- whatever amount of exceeding ../ the root container URL has to remain itself ("vertical" sense) as per specs
2- ../ parts combined with path segments can lead outside of the container in "horizontal" sense, like in a sibling folder of the root folder both with an exact number of ../ and with an exceeding number of ../

So we can say that type 1 is allowed, while type 2 is forbidden.
The algorithm seems to be designed to detect both, if I am not wrong, but this relies on the parsing, that "stops" at example.org, not at A, I think.

-About the root container URL, it has to be created, but it is just that / is enough when there is not a folder like mobydick/ ?
Or it is a complete URL?

Thank you

**Comment by mattgarrish** [comment]
> If I put this URL `https://url.spec.whatwg.org/https://github.com/w3c/epub-specs/issues/1948` in the browser address bar I get "404 Not Found" error. It does not become
> 
> `https://github.com/w3c/epub-specs/issues/1948`

I think you're missing that this algorithm does not apply to absolute url strings. You have to go up a bit from the start of the algorithm but the section says:

> In the OCF Abstract Container , when a file uses a URL string to reference another file in the container, the string MUST be a path-relative-scheme-less-URL string , optionally followed by U+0023 (#) and a URL-fragment string.

Could the algorithm not also say "For a path-relative-scheme-less-URL string..." instead of "For a URL string"? I know it's wordier and has to be true by the earlier prose, but if it avoids confusion... Or perhaps the explanatory paragraph could be moved down closer to the start of the algorithm? (the note after it doesn't appear to follow directly from anything in that paragraph, at least)

**Comment by iherman** [comment]
> @rdeltour
> 
> -Sorry I cannot find explanation or examples in that official document to understand how the URL is also pieces of URL. Any direct link to part of the documents would be welcome.
> 
> -If I put this URL `https://url.spec.whatwg.org/https://github.com/w3c/epub-specs/issues/1948` in the browser address bar I get "404 Not Found" error. It does not become
> 
> https://github.com/w3c/epub-specs/issues/1948
> It is not accepted or parsed. I also tried `https://url.spec.whatwg.org/A/https://github.com/w3c/epub-specs/issues/1948` with the same results. There is something I do not grab here. Can you explain? What URL parsers do you mean? Where is the official rule in the document to do so?

To be honest, I do not really understand how you came to these URLs, and where the "URL is also pieces of URL" comes from.

I think that you should really read the official parser algorithm that is described here:

https://url.spec.whatwg.org

I realize this is not an easy read (to say the least), but that is the nature of the beast. There are some additional description that might be helpful here:

https://developer.mozilla.org/en-US/docs/Web/API/URL_API

There is also a "playground":

https://jsdom.github.io/whatwg-url/

that you can use for specific URLs. For example, your URL above yields:

https://jsdom.github.io/whatwg-url/#url=aHR0cHM6Ly91cmwuc3BlYy53aGF0d2cub3JnL2h0dHBzOi8vZ2l0aHViLmNvbS93M2MvZXB1Yi1zcGVjcy9pc3N1ZXMvMTk0OA==&base=YWJvdXQ6Ymxhbms=

which also shows that there is no such thing as a "URL is also pieces of URL": what _looks_ like an (embedded) URL is not considered as such but, instead, is a path.

**Comment by rdeltour** [comment]
@mattgarrish
> I think you're missing that this algorithm does not apply to absolute url strings.

In fact it does 😊. It applies to any URL string. Step 9 make absolute URL strings pass.

Of course, a checker implementation would probably only apply it to relative URL strings. But making it pass for any string was needed anyway to take overridden base URLs into account (for instance, the relative URL string "`TR/epub-33/`" in an HTML hyperlink, with a `base` element set to "`https://www.w3.org`").

@P5music 
> -Sorry I cannot find explanation or examples in that official document to understand how the URL is also pieces of URL.
> Any direct link to part of the documents would be welcome.

I just see @iherman provided links. I second the recommendation to read them.

As for the formal definitions (referred to in EPUB):
- [URL](https://url.spec.whatwg.org/#concept-url) (and its components)
- [valid URL string](https://url.spec.whatwg.org/#valid-url-string) (and sub-categories)
- [URL parser](https://url.spec.whatwg.org/#concept-url-parser)

> -If I put this URL
> https://url.spec.whatwg.org/https://github.com/w3c/epub-specs/issues/1948
> in the browser address bar
> I get "404 Not Found" error.
> 
> It does not become
> 
> https://github.com/w3c/epub-specs/issues/1948
> It is not accepted or parsed.


Like @iherman, I don't see how you try this URL string. My guess is that you combine a base URL with a URL string by concatenating them, but as I said earlier, this is not how this works.
A base URL is combined with a URL string when the URL string is parsed. Web browsers apply the parsing algorithm whenever they encounter or process a URL string.

If you open an HTML document that has a single element `base` with an `href` attribute set to `https://url.spec.whatwg.org/`, and this document has a hyperlink `a` element with an `href` attribute set to `https://github.com/w3c/epub-specs/issues/1948`, the browser will parse this latter value with the base, and return `https://github.com/w3c/epub-specs/issues/1948`.

> Can you explain? What URL parsers do you mean? Where is the official rule in the document to do so?

This is all [defined in HTML](https://html.spec.whatwg.org/#resolving-urls), which transitively refers to the URL parser algorithm defined in the URL standard (linked above).

We provided links to the standard we're discussing everywhere, from the very first issue comment in #1888, as well as in all specification PRs. I'm surprised by the question "What URL parsers do you mean?" when we've been talking about this and provided hyperlinks for weeks now!

> -About the two kind of leaking URLs, they are
> out-of-container URLs
> and
> going-too-much-up URLs
> They are not the same because there are two kind of problems with URLs for the ePub3:
> 
> 1- whatever amount of exceeding ../ the root container URL has to remain itself ("vertical" sense) as per specs
> 2- ../ parts combined with path segments can lead outside of the container in "horizontal" sense, like in a sibling folder of the root folder both with an exact number of ../ and with an exceeding number of ../
>
> So we can say that type 1 is allowed, while type 2 is forbidden.

I still don't see the difference. A concrete example would help. 

> -About the root container URL, it has to be created, but it is just that / is enough when there is not a folder like mobydick/ ?
> Or it is a complete URL?

It is a complete URL.
`epub://mobydick-12345/` would (kinda) work. `http://localhost:12345/` would work.
`http://localhost:12345/mobydick` would not.

`/` only is not a URL (more precisely: it is not a valid absolute URL string, so it is not the serialization of a URL).

**Comment by mattgarrish** [comment]
> Step 9 make absolute URL strings pass.

Okay, I see that now (guilty of skimming), but that's a confusing transition into the algorithm, then. You're saying they must be relative to resolve within the container, then there's an algorithm that doesn't really say why it's there except that URLs have to return true at the end.

I'm not quite sure what the exact name should be, but it would help if it were a little more self-explanatory from the outset. For example:

> To **validate a URL** in the OCF Container given a URL string _url_, ...

**Comment by P5music** [comment]
@rdeltour

Please can you explain where respectively url means:
-a relative url (a piece of URL)
-a url record (an entire URL)

in regard to:
the content context,
the base URL,
the root container URL,
urls that are found in XHTML documents, and so on if other cases exist.

I am asking this because,
it seems to me (so I could be wrong) that
the specification text is not being enough accurate when using that terms 
(it is just a technical evaluation, no offence intended).

I remind that you created the algorithm just because you were dealing with relative urls, and path segments, 
that seem to constitute the ePub publication very structure.

Now the specification seems to be mandating some reasonable constraints but
it seems that sometimes one meaning (relative url) is used or implied, sometimes the other (url record) is used or implied.

I am advocating the reader's point of view, or the RSs' point of view.

The main consequence of the above uncertainty or weakness is that
it is not so clear that it refers to urls that are already formed before checking them,
while still appearing that you are explaining how to concatenate the path segments.
Or this is somehow what the reader could expect.

You in fact said above that you are not using an explicit concatenation concept, so you are saying that the url are formed "their way",
except that you want they comply with some constraints.

The main constraints I can see are the
1-going too much up URLs (vertical)
2-leaking URLs in sibling folder (horizontal)

The algorithm seems to be accounting for both, now that I understand the URL parsing rules, but they are somehow unexpected for the reader.

Also the two types of root container URL the spec deals with 
now drop the root folder concept that was in force before, or at least in the mind of people dealing with the ePub3 archive.
It's a big step that should be thoroughly explained and evident.

When the text is revised I could also check for some bugs in it, but now it is just generating misunderstanding IMHO.

Regards

**Comment by iherman** [comment]
@P5music,

> Please can you explain where respectively url means: 
> - a relative url (a piece of URL) 
> - a url record (an entire URL)
> 
> in regard to: 
> - the content context, 
> - the base URL, 
> - the root container URL, 
> - urls that are found in XHTML documents, and so on if other cases exist.
> 
> I am asking this because, it seems to me (so I could be wrong) that the specification text is not being enough accurate when using that terms (it is just a technical evaluation, no offence intended).

you should be more specific about what your questions are. Those terms are (and should) be specified either in the document or in the URL spec and, without specific questions, I would not know what to answer.

**Comment by iherman** [comment]
> 
> I'm not quite sure what the exact name should be, but it would help if it were a little more self-explanatory from the outset. For example:
> 
> > To **validate a URL** in the OCF Container given a URL string _url_, ...

I like that change, so I went ahead and did it... (per https://github.com/w3c/epub-specs/pull/1939/commits/587f2b9e41667bf838e0a914e996eb20714bd87e).

**Comment by rdeltour** [comment]
> Please can you explain where respectively url means: -a relative url (a piece of URL) -a url record (an entire URL)

The terms **always** mean what they are defined to mean by the URL standard.

> I am asking this because, it seems to me (so I could be wrong) that the specification text is not being enough accurate when using that terms (it is just a technical evaluation, no offence intended).

Can you point at a concrete example where the spec is not enough accurate?

> Now the specification seems to be mandating some reasonable constraints but it seems that sometimes one meaning (relative url) is used or implied, sometimes the other (url record) is used or implied.

Again, the specification use the terms defined by the  URL standard.
If you think the specification is wrong in some place (that can happen!), or not explicit enough, please provide the exact excerpt were you think it is.

> The main consequence of the above uncertainty or weakness is that it is not so clear that it refers to urls that are already formed before checking them, while still appearing that you are explaining how to concatenate the path segments. Or this is somehow what the reader could expect.
> 
> You in fact said above that you are not using an explicit concatenation concept, so you are saying that the url are formed "their way", except that you want they comply with some constraints.

Have you read the URL parser algorithm? If you don't understand it, it's OK. It is hard to read.
It's not me saying things, it is the standard EPUB normatively references.

> The main constraints I can see are the 1-going too much up URLs (vertical) 2-leaking URLs in sibling folder (horizontal)

I still don't understand what these mean, without a concrete example.

> The algorithm seems to be accounting for both, now that I understand the URL parsing rules, but they are somehow unexpected for the reader.
> 
> Also the two types of root container URL the spec deals with now drop the root folder concept that was in force before, or at least in the mind of people dealing with the ePub3 archive. It's a big step that should be thoroughly explained and evident.
> 
> When the text is revised I could also check for some bugs in it, but now it is just generating misunderstanding IMHO.

So far only one person (you) expressed disagreement due to misunderstanding.

Reading standards is admittedly difficult. It takes some time to understand definitions and logic on how the Web works, even for common concepts like URLs. That complexity is the cost of precise and unambiguous specification.

Some notes or explainers might help. But without any concrete example of what you don't understand, we cannot explain anything more that what is (proposed) in the spec.

**Comment by P5music** [comment]
@iherman
@rdeltour

That's the best form I can provide for my concerns.
Indeed it is not useful to check every single line of that spec part here
because
you have to frame the entire picture.

I can just contribute with further elements to support my opinion:
Let's start from the beginning
>The properties of the <a>container root URL</a> are such that whatever the amount
						of <a data-cite="url/#double-dot-path-segment">double-dot path segments</a> in a URL string (for
						example, <code>../../../secret</code>), it is parsed to a <a>content URL</a> (and not "leak"
						outside the container). This avoids potential run-time security issues. 
						Furthermore, the additional constraint and <a href="#algo-out-of-container">algorithm</a> below 
						ensure that such potentially problematic URLs can also be detected when checking the EPUB Document.


In this form, it seems that this is a definition of the container root URL,
not a constraint for the urls inside the container.
It is just like to say:
"be the container root URL the URL that is at the highest level relative to the contained urls so they do not exceed".
One could be tricked into creating, for example, a container-root-URL checker, an application that calculates the container root URL for a given ePub, maybe in a place where many ePubs reside, 
and if a wrong url with many ../.. is present in any of the document files of that ePub,
the resulting calculated container root URL will be a very "high" root.
What I am saying clearly makes not sense because it is an extreme example.
But I just am pointing out that it seems a constraint for the container root URL, not for the urls.

But this also has some real implications.
Let this be an example:
http://myepubwebsite/epubs/literature/novels/mobydick
This reflects the way a website is organized.
I was said that there are possible cases where not just a simple filesystem organization has to be taken into account,
up to systems that in principle serve the ePubs directly.

In those cases a container root URL has to be created, as it is in the specs,
but it could not reflect the categorization like in the examples, it could be a special ID like it is in some "StorageAccessFramework" or other cloud systems.
And also in the case the url path reflects the categorization, a different categorization could be possible, so that the same website could find the ePub at another URL, if I am not wrong.
For example
http://myepubwebsite/epubs/literature/authors/melville/mobydick
could be also valid.

Now, the specs clearly deal with ../../../ path segments, in the relative URLs
that are internal to the ePub
but we see that outside the ePub "real container root" other path segments could be inserted, 
also with ../.. possible wrong parts or exceeding or leaking.

So I feel that the "real container root" is something it is better not to drop as a fundative concept,
especially when transitioning to URL complying mandatory constraints.

Do you agree with some of my points?
Does this make you discuss more the specs?

Furthermore I think that using obscure URL parsing rules nobody reminds when reading the specs for the container root URL for the ePub, 
especially when it is to support an algorithm that is suggested,
and that the reader has to fathom in his/her mind to figure out the whole thing, is not convenient.

These are only suggestions IMHO.

Thanks
Regards

**Comment by rdeltour** [comment]
> That's the best form I can provide for my concerns. Indeed it is not useful to check every single line of that spec part here because you have to frame the entire picture.

It would be easier for us to start from something concrete and specific in the proposal, rather than having to do some guesswork on your understanding of the entire picture.

> 
> I can just contribute with further elements to support my opinion: Let's start from the beginning
> 
> > The properties of the container root URL are such that whatever the amount
> > of double-dot path segments in a URL string (for
> > example, `../../../secret`), it is parsed to a content URL (and not "leak"
> > outside the container). This avoids potential run-time security issues.
> > Furthermore, the additional constraint and [algorithm](#algo-out-of-container) below
> > ensure that such potentially problematic URLs can also be detected when checking the EPUB Document.
> 
> In this form, it seems that this is a definition of the container root URL, not a constraint for the urls inside the container

No, it is not a definition, it is an informative note.
It says that **given** the normative properties of the container root URL, even URL strings with more `..` than needed will be resolved to something inside the container after parsing.

> It is just like to say: "be the container root URL the URL that is at the highest level relative to the contained urls so they do not exceed". 

No, I think you understand it wrong.
The container root URL **does not** depend on the URL strings in the EPUB; it is defined (internally) by the reading system. It only has to follow the given properties. The reading system does not need to know or list all the URL strings present in the EPUB to compute the URL.

To repeat the example listed in the spec, `http://localhost:49152/` is a conforming container root URL.

> One could be tricked into creating, for example, a container-root-URL checker, an application that calculates the container root URL for a given ePub, maybe in a place where many ePubs reside, and if a wrong url with many ../.. is present in any of the document files of that ePub, the resulting calculated container root URL will be a very "high" root.

No, again you cannot "calculate the container root URL for a given EPUB". The root URL is not calculated or computed, it is arbitrarily set by the reading system, following their own implementation logic, which is conforming as long as it respects the given properties.

> What I am saying clearly makes not sense because it is an extreme example.

An actual concrete example would help.

> But I just am pointing out that it seems a constraint for the container root URL, not for the urls.
> 
> But this also has some real implications. Let this be an example: http://myepubwebsite/epubs/literature/novels/mobydick 

It seems you conflate the URL of the publication (the actual EPUB) with the URL of the container root. The URL above is not a conforming container root URL.

> This reflects the way a website is organized. I was said that there are possible cases where not just a simple filesystem organization has to be taken into account, up to systems that in principle serve the ePubs directly.
> 
> In those cases a container root URL has to be created, as it is in the specs, but it could not reflect the categorization like in the examples, it could be a special ID like it is in some "StorageAccessFramework" or other cloud systems.

I don't understand what "it could be a special ID like it is in some "StorageAccessFramework" or other cloud systems" means. Please, please, use concrete examples. Do not assume people know what "special ID" means, nor "StorageAccesFramework", nor "cloud systems". This makes it very hard to follow your reasoning.

> And also in the case the url path reflects the categorization, a different categorization could be possible, so that the same website could find the ePub at another URL, if I am not wrong. For example http://myepubwebsite/epubs/literature/authors/melville/mobydick could be also valid.

A valid URL to the publication, maybe. A valid container root URL (used internally by the reading system), no.


> Now, the specs clearly deal with ../../../ path segments, in the relative URLs that are internal to the ePub but we see that outside the ePub "real container root" other path segments could be inserted, also with ../.. possible wrong parts or exceeding or leaking.

I cannot make any sense of that. What do you mean by "other path segments could be inserted"?

> 
> So I feel that the "real container root" is something it is better not to drop as a fundative concept, especially when transitioning to URL complying mandatory constraints.
> 
> Do you agree with some of my points?

I can't say I agree (not really that I disagree for that matters). Mostly, I don't understand your points for lack of examples, or because I don't understand the terms or concepts you're using without definition.

> Does this make you discuss more the specs?

Not really. 
If you want to discuss it further, start with a concrete piece of the specification. Give a concrete (fully described) example. Explain what you understand the specification means for your example, and how you think it is wrong.

If you cannot do that, I'm sorry but I won't be further trying to guess your points.

> Furthermore I think that using obscure URL parsing rules nobody reminds when reading the specs for the container root URL for the ePub, especially when it is to support an algorithm that is suggested, and that the reader has to fathom in his/her mind to figure out the whole thing, is not convenient.

I wouldn't call the URL parser an "obscure URL parsing rule", given it's a foundational piece of the Web that billion users rely on everyday.

As for the other algorithm to discard "out-of-container" URL strings in EPUB, it's not critical to understand (SHOULD), and the notes, I believe, explain what it means well-enough.

**Comment by iherman** [comment]
> I can just contribute with further elements to support my opinion: Let's start from the beginning
> 
> > The properties of the container root URL are such that whatever the amount
> > of double-dot path segments in a URL string (for
> > example, `../../../secret`), it is parsed to a content URL (and not "leak"
> > outside the container). This avoids potential run-time security issues.
> > Furthermore, the additional constraint and [algorithm](#algo-out-of-container) below
> > ensure that such potentially problematic URLs can also be detected when checking the EPUB Document.
> 
> In this form, it seems that this is a definition of the container root URL, not a constraint for the urls inside the container.

I respectfully disagree. We try to be fairly meticulous on the spec: the term "container root URL" is an active link to the terminology section, ie, this is _**not**_ a definition. This note should be looked in the context of the whole specification which has some strict typesetting conventions that it follows.

**Comment by P5music** [comment]
@iherman
@rdeltour

The fact is that you are creating constraints for things that are extraneous to the ePub3 publications per se.

I understand some of your concerns about URL standard and other important aspects, 
but I really think they should not find place in the ePub3 specs.
You are bringing webserver behaviours, URL standard subtleties, internet common practices, URL parsing edge cases
into
the ePub3 standard specifications.
I'd like to read other people's opinions, just to understand what I am so wrong.

Regards

**Comment by rdeltour** [comment]
> The fact is that you are creating constraints for things that are extraneous to the ePub3 publications per se.

I still don't understand why you think so. Granted, EPUB RS spec creates constraints on Reading Systems, which are "extraneous to the ePub3 publications per se". But beyond that, I cannot see.

> I understand some of your concerns about URL standard and other important aspects, but I really think they should not find place in the ePub3 specs. You are bringing webserver behaviours, URL standard subtleties, internet common practices, URL parsing edge cases into the ePub3 standard specifications. 

Sigh. 
**I** am not bringing URL standard subtleties, or URL parsing edge cases, into the EPUB standard. The **working group** did when it decided to use URLs in EPUB, i.e. when EPUB was created.

The fact that the WG recently decided to normatively reference the URL standard rather than the old deprecated RFC doesn't change much. The URL standard is, admittedly, more difficult to read if you're not used to modern standards and algorithmic specifications, but it is also more precise and harmonized with Web browser implementations, so that's a big benefit in standardisation space.

It **is** a fact that when a standard normatively references another standard, it does bring all its subtleties and edge cases. It's always for the better, since it does clarify the edge cases for the implementors.

**Comment by wareid** [comment]
@P5music I think Romain and Ivan have likely made all of the points we can make on this topic, but just to clarify a few things. With these latest revisions we are bringing EPUB more closely into the larger web technologies ecosystem, which both makes some things easier and others more complex. Referencing specifications like the URL standard or HTML and then specifying within EPUB our relationship to those is the practice we've decided on as a WG. This of course means we are at times clarifying or simplifying things from the EPUB spec, and introducing the complexities of other specs into our realm. 

I am going to propose we close this issue as I believe we've exhausted the discussion on this, but if you disagree, I just ask that we focus discussion on matters the EPUB WG can undertake, which means any issues relating to the URL standard or other external specifications are out of scope for this discussion. If there are matters here that have impact on RS development or EPUB development, please help us with concrete examples of issues so we can discuss managing them.

**Comment by P5music** [comment]
@wareid

I am sorry you feel the urge to close this discussion, 
it would be a good thing to let other people say something 
especially because I explicitly asked for that.

I pointed out something relevant in other threads at the point that it ended up being written into the specs as important aspects.

Nobody else raised those precious objections.

I also see that

the URL standard and other things have been mentioned above
just apply now!

It is not necessary to remind people dealing with ePub3 about the mentioned subtleties.

They are naturally enforced when real implementations come along.

Nobody doubt that the ePub3 is entwined with W3 technology, because it stemmed from it. 
I am saying that from my first contribution here.

Some of the concerns are reasonable and they are going to be transformed into constraints. I agree.
I think they are very easy to express in prose, but this opinion has already been discussed.

My proposal is to normate just what is not already said in the W3C standard. 

So if the RSs have to do something then tell them to do,
but do not say them how to handle URLs because it is their job,
even if those URLs are constructed from the relative URLs that are internal to ePub3 publications.

If the ePub3 is normated within its real perimeter, it is compatible with both webservers and filesystems.

ePub3 has relative URLs inside, that's it. Put some constraints on them, it's OK.

Regards

**Comment by rdeltour** [comment]
> My proposal is to normate just what is not already said in the W3C standard.
> 
> So if the RSs have to do something then tell them to do, but do not say them how to handle URLs because it is their job, even if those URLs are constructed from the relative URLs that are internal to ePub3 publications.
> (…)
> ePub3 has relative URLs inside, that's it. Put some constraints on them, it's OK.

It is unfortunate that you do not understand it, but this is exactly what we're doing! 😊

**Comment by P5music** [comment]
@rdeltour

I show you what I think you are doing according to my humble interpretation:

I know that special web technologies exist nowadays that deal with URL that are not in the common path-segments form.

Cloud providers for example create URLs in dynamic way, the same document can change its URL, but it stays in the same position in some secret vault.
Even when the URL is static it is in a special form, an ID is assigned, and the URL has special form also before that ID.
You cannot understand what is going on there.

Those URLs can have special form, for example in Android StorageAccessProvider or Google or Apple cloud systems there is something like
https://drive.google.com/file/d/adasdsHuwDRt4FP98523vn2987cn22v5

This is not a real URL for a real file, I changed the characters,
but let's say I started from one of my copies of EPUB_3.epub book I have in my cloud folders.

You are using the ../.. segments concept to constrain the relative URLs inside the ePub archive (OK)
and also outside the ePub archive (not OK in this case).
You imagine your algorithm being appliable, 
step by step
hatching the egg of the ePub3, going up with those ../.

It's wrong. The algorithm is wrong, it cannot be applied.

I hope this example is useful.
Feel free to ask me for further explanation.
Regards

**Comment by dauwhe** [comment]
To better understand how URLs are parsed, I would suggest playing around with https://jsdom.github.io/whatwg-url/

**Comment by rdeltour** [comment]
@P5music 

> You are using the ../.. segments concept to constrain the relative URLs inside the ePub archive (OK)
> and also outside the ePub archive (not OK in this case).
> You imagine your algorithm being appliable,
> step by step
> hatching the egg of the ePub3, going up with those ../.

I really do not understand what you mean at all, sorry. I'll repeat one last time what I said earlier:

If you want to discuss it further:
- point to a concrete piece of the specification (you can quote the part you'd like to discuss)
- give a concrete (fully described) example: if possible, the file and directory structure of the EPUB, and markup examples where URL strings appear.
- explain what you understand the specification means for your example, and why you think it is wrong.


> Feel free to ask me for further explanation.

I keep asking for concrete examples of where the proposed changes are bad. Unless you do the above, I'll stop now, it's frustrating and time consuming for all of us.
